### PR TITLE
feat!: dont batch queries

### PR DESCRIPTION
**Pull request by samsiegart**
refs https://github.com/Agoric/agoric-sdk/issues/8505

We've already switched from using the batch RPC API to JSON REST API in https://github.com/Agoric/ui-kit/pull/55. But, we were still using a `Promise.all()` with the new interface, effectively batching those queries together. This PR makes each vstorage path poll asynchronously on separate timers.

Importantly, this changes the error handling to be more manageable. Instead of having a separate error handler for each `watchLatest`, it just uses a single `onError` callback given to `chainStorageWatcher` constructor. The new JSON API doesn't include an optional error message+code in response objects--it's limited to just http-level error codes, and in the case of missing vstorage data, an empty `value` in the response. This allows us to use axios to automatically retry on 500-level errors a couple times before invoking `onError`.

This updated error handling allows for clients to more easily manage faulty RPC nodes. Instead of adding an error handler to every place `watchLatest` is used, and use the result to decipher whether the vstorage data is missing or the RPC node is failing, it can just use the single `onError` callback on the `chainStorageWatcher`, and trust that the `chainStorageWatcher` already retried a couple times. This gives the client a reliable way to know when to notify the user of an RPC outage and/or prompt them to select a different node (assuming a node selector is available).

For `queryOnce`, it also no longer batches with other requests, and uses axios to retry before throwing an error, more similar to a regular `fetch` request.

Added some test cases for the updated error handling.

### CODE REVIEW

File: packages/rpc/test/chainStorageWatcher.test.ts [hunk 1]
```diff
@@ -224,22 +350,23 @@ const createFetchResponse = (
...
-          : {
-              value: JSON.stringify({
-                values: json ? [JSON.stringify(marshal(value))] : undefined,
-                value: !json ? value : undefined,
-                blockHeight: String(blockHeight ?? 0),
-              }),
+  code = 200,
+) =>
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  new Promise<any[]>(res => {
```
[the remark this answers is not in the document]
Comment (samsiegart) [reply packages/rpc/test/chainStorageWatcher.test.ts]: Hm now it does, just changed. Before I was getting some issue with 'axios-mock-adapter' expecting `Promise<any>`

File: packages/rpc/src/vstorageQuery.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+axiosRetry(axiosClient, {
+  retries: 2,
+  retryDelay: axiosRetry.exponentialDelay,
+});
+
+export const vstorageQuery = async (
+  node: string,
+  unmarshal: FromCapData<string>,
+  path: [AgoricChainStoragePathKind, string],
+) => {
```
[the remark this answers is not in the document]
Comment (samsiegart) [reply packages/rpc/src/vstorageQuery.ts]: Done (except had to make `value` a generic and the return value a `Promise`)

File: packages/rpc/src/chainStorageWatcher.ts [hunk 3]
```diff
@@ -195,34 +154,28 @@ export const makeAgoricChainStorageWatcher = (
...
+      path,
+    );
+    return value;
+  };
 
   // Assumes argument is an unserialized presence.
   const presenceToSlot = (o: unknown) => marshaller.toCapData(o).slots[0];
 
-  const queryBoardAux = <T>(boardObjects: unknown[]) => {
+  const queryBoardAux = (boardObjects: unknown[]) => {
```
[the remark this answers is not in the document]
Comment (samsiegart) [reply packages/rpc/src/chainStorageWatcher.ts]: Just added back, but I had to make `vstorageQuery` generic as well

File: packages/rpc/src/chainStorageWatcher.ts [hunk 4]
```diff
@@ -67,120 +65,81 @@ export const makeAgoricChainStorageWatcher = (
...
+      (blockHeight === lastValue[0] ||
+        (blockHeight === undefined && JSON.stringify(value) === lastValue[0]))
+    ) {
+      // The value isn't new, don't emit.
+      queueNextRefresh();
+      return;
     }
+
+    latestValueCache.set(pathKey, [
+      blockHeight ?? JSON.stringify(value),
```
[the remark this answers is not in the document]
Comment (samsiegart) [reply packages/rpc/src/chainStorageWatcher.ts:121]: Done

File: packages/rpc/src/chainStorageWatcher.ts [hunk 5]
```diff
@@ -67,120 +65,81 @@ export const makeAgoricChainStorageWatcher = (
...
-          s.onUpdate(harden(value));
-        });
-      });
-    } catch (e) {
-      onError && onError(e as Error);
-    } finally {
-      isQueryInProgress = false;
-      queueNextQuery();
+    const { value, blockHeight } = response;
+    const lastValue = latestValueCache.get(pathKey);
```
[the remark this answers is not in the document]
Comment (samsiegart) [reply packages/rpc/src/chainStorageWatcher.ts]: I just made a type for `LatestValueIdentifier` with a doc comment to hopefully clarify, and used the array destructuring you suggested

File: packages/rpc/src/chainStorageWatcher.ts [hunk 6]
```diff
@@ -51,136 +58,101 @@ export const makeAgoricChainStorageWatcher = (
...
-      isQueryInProgress = false;
+    const pathKey = pathToKey(path);
+    let response;
+    try {
+      response = await vstorageQuery(apiAddr, marshaller.fromCapData, path);
+    } catch (e: unknown) {
+      console.error(`Error querying vstorage for path ${path}:`, e);
+      if (e instanceof Error) {
+        onError && onError(e);
+      }
```
Comment (turadg) [inline packages/rpc/src/chainStorageWatcher.ts:96]: consider one conditional expression,
```suggestion
      if (onError && e instanceof Error) {
        onError(e);
      }
```

File: packages/rpc/src/chainStorageWatcher.ts [hunk 7]
```diff
@@ -51,136 +58,101 @@ export const makeAgoricChainStorageWatcher = (
...
-      queueNextQuery();
+    const { value, blockHeight } = response;
+    const [latestValueIdentifier, latestValue] =
+      latestValueCache.get(pathKey) || [];
+
+    if (
+      latestValue &&
+      (blockHeight === latestValueIdentifier ||
+        // Blockheight is undefined so fallback to using the stringified value
+        // as the identifier, as is the case for `children` queries.
```
Comment (turadg) [inline packages/rpc/src/chainStorageWatcher.ts:110]: helpful comment

File: packages/rpc/src/vstorageQuery.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+axiosRetry(axiosClient, {
+  retries: 2,
+  retryDelay: axiosRetry.exponentialDelay,
+});
+
+export const vstorageQuery = async <T>(
+  node: string,
+  unmarshal: FromCapData<string>,
+  path: [AgoricChainStoragePathKind, string],
```
Comment (turadg) [inline packages/rpc/src/vstorageQuery.ts:39]: consider naming these so they're clearer below
```suggestion
 [pathKind, path]: [AgoricChainStoragePathKind, string],
```

File: packages/rpc/src/vstorageQuery.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        value: res.value,
+      };
+    }
+
+    const data = parseIfJSON(res.value);
+
+    const latestValue =
+      typeof data.values !== 'undefined'
+        ? parseIfJSON(data.values[data.values.length - 1])
+        : parseIfJSON(data.value);
```
Comment (turadg) [inline packages/rpc/src/vstorageQuery.ts:60]: superfluous negation makes it harder to read
```suggestion
    const latestValue =
      typeof data.values === 'undefined'
        ? parseIfJSON(data.value);
        : parseIfJSON(data.values[data.values.length - 1])
```

ditto below

**Review by turadg — APPROVED** [review]

File: packages/rpc/src/vstorageQuery.ts (hunk 9 above)
Comment (samsiegart) [reply packages/rpc/src/vstorageQuery.ts]: Done here and below

File: packages/rpc/src/vstorageQuery.ts (hunk 8 above)
Comment (samsiegart) [reply packages/rpc/src/vstorageQuery.ts]: Done

File: packages/rpc/src/chainStorageWatcher.ts (hunk 6 above)
Comment (samsiegart) [reply packages/rpc/src/chainStorageWatcher.ts]: Done

**Pull request merged**